### PR TITLE
Groupby Agg not working if different partials with same name on the same column

### LINKED ISSUE #28570: Groupby Agg not working if different partials with same name on the same column
related to #28426  from PR #28428 

from functools import partial

import pandas as pd
import numpy as np

quant50 = partial(np.percentile, q=50)
quant70 = partial(np.percentile, q=70)

test = pd.DataFrame({'col1': ['a', 'a', 'b', 'b', 'b'], 'col2': [1,2,3,4,5]})
test.groupby('col1').agg({'col2': [quant50, quant70]})

However, quant50 result is rewritten by quant70, so will get the same output on both columns.

output:

| col1  |  percentile | percentile  |
|---|---|---|
|  a | 1.7  |  1.7 |
|  b | 4.4  |  4.4 |

expected output:

| col1  |  percentile | percentile  |
|---|---|---|
|  a | 1.5  |  1.7 |
|  b | 4.0  |  4.4 |

### PR DESCRIPTION
**Pull request by aliraeisdanaei**
- [x] closes #28570
- [x] tests added / passed
- [x] Ensure all linting tests pass, see [here](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit) for how to run them
- [x] whatsnew entry
a unit test was added in accordance with issue [28570](https://github.com/pandas-dev/pandas/issues/28570). It asserts the equality of the expected output with the tested df.

### CODE REVIEW

**Comment by aliraeisdanaei** [comment]
Hey, this is my first pull request to pandas, so please do tell me if I have made any mistakes. 

Thank you & God bless,
Ali Raeisdanaei

File: pandas/tests/groupby/aggregate/test_aggregate.py [hunk 1]
```diff
@@ -62,6 +62,33 @@ def test_agg_ser_multi_key(df):
...
+        if not inplace:
+            df = df.copy()
+        df.columns = pd.MultiIndex.from_product([[top_col], df.columns])
+        return df
+    quant50 = partial(np.percentile, q=50)
+    quant50.__name__ = "quant50"
+    quant70 = partial(np.percentile, q=70)
+    quant70.__name__ = "quant70"
+
+    test = pd.DataFrame({'col1': ['a', 'a', 'b', 'b', 'b'], 'col2': [1,2,3,4,5]})
```
Comment (jbrockmendel) [inline pandas/tests/groupby/aggregate/test_aggregate.py:78]: see the doc on code_style; running `black` should take care of this

File: pandas/tests/groupby/aggregate/test_aggregate.py [hunk 2]
```diff
@@ -62,6 +62,33 @@ def test_agg_ser_multi_key(df):
...
+    test = test.groupby('col1').agg({'col2': [quant50, quant70]})
+    # print(test)
```
Comment (jbrockmendel) [inline pandas/tests/groupby/aggregate/test_aggregate.py:80]: please remove commented-out

File: pandas/tests/groupby/aggregate/test_aggregate.py [hunk 3]
```diff
@@ -62,6 +62,33 @@ def test_agg_ser_multi_key(df):
...
+
+    expected_df = pd.DataFrame({'col1': ['a', 'b'],
+                                'quant50': [1.5, 4.0],
+                                'quant70': [1.7, 4.4]})
+    expected_df = expected_df.set_index('col1')
+    expected_df = add_top_column(expected_df, 'col2')
+    # print(expected_df)
+
+    assert(test.equals(expected_df))
```
Comment (jbrockmendel) [inline pandas/tests/groupby/aggregate/test_aggregate.py:89]: use tm.assert_frame_equal(test, expected_df)

**Comment by aliraeisdanaei** [comment]
@jbrockmendel Thanks for the feedback.

File: pandas/tests/groupby/aggregate/test_aggregate.py [hunk 4]
```diff
@@ -62,6 +62,32 @@ def test_agg_ser_multi_key(df):
     tm.assert_series_equal(results, expected)
 
 
+def test_agg_different_partials():
+    # for issue 28570
+
+    def add_top_column(df, top_col, inplace=False):
```
Comment (mroeschke) [inline pandas/tests/groupby/aggregate/test_aggregate.py:68]: Looks like this function is only used once. Could you in-line the logic directly in the test?

Comment (mroeschke) [reply pandas/tests/groupby/aggregate/test_aggregate.py]: Rather, this _entire function_ is only used once so all the logic can be in-lined.

File: pandas/tests/groupby/aggregate/test_aggregate.py [hunk 5]
```diff
@@ -62,6 +62,30 @@ def test_agg_ser_multi_key(df):
...
+    quant70.__name__ = "quant70"
+
+    test = pd.DataFrame({"col1": ["a", "a", "b", "b", "b"], "col2": [1, 2, 3, 4, 5]})
+    test = test.groupby("col1").agg({"col2": [quant50, quant70]})
+
+    expected_df = pd.DataFrame(
+        {"col1": ["a", "b"], "quant50": [1.5, 4.0], "quant70": [1.7, 4.4]}
+    )
+    expected_df = expected_df.set_index("col1")
+    expected_df = add_top_column(expected_df, "col2")
```
Comment (jreback) [inline pandas/tests/groupby/aggregate/test_aggregate.py:84]: can you just inline the add_top_column, as a function just obfuscates the purpose

File: pandas/tests/groupby/aggregate/test_aggregate.py [hunk 6]
```diff
@@ -62,6 +62,30 @@ def test_agg_ser_multi_key(df):
...
+    def add_top_column(df, top_col):
+        df.columns = pd.MultiIndex.from_product([[top_col], df.columns])
+        return df
+
+    quant50 = partial(np.percentile, q=50)
+    quant50.__name__ = "quant50"
+    quant70 = partial(np.percentile, q=70)
+    quant70.__name__ = "quant70"
+
+    test = pd.DataFrame({"col1": ["a", "a", "b", "b", "b"], "col2": [1, 2, 3, 4, 5]})
```
Comment (jreback) [inline pandas/tests/groupby/aggregate/test_aggregate.py:77]: call this df

File: pandas/tests/groupby/aggregate/test_aggregate.py [hunk 7]
```diff
@@ -62,6 +62,30 @@ def test_agg_ser_multi_key(df):
...
+    test = test.groupby("col1").agg({"col2": [quant50, quant70]})
```
Comment (jreback) [inline pandas/tests/groupby/aggregate/test_aggregate.py:78]: call this result

File: pandas/tests/groupby/aggregate/test_aggregate.py [hunk 8]
```diff
@@ -62,6 +62,30 @@ def test_agg_ser_multi_key(df):
...
+
+    expected_df = pd.DataFrame(
```
Comment (jreback) [inline pandas/tests/groupby/aggregate/test_aggregate.py:80]: just call this expected

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/tests/groupby/aggregate/test_aggregate.py [hunk 9]
```diff
@@ -62,6 +62,26 @@ def test_agg_ser_multi_key(df):
     tm.assert_series_equal(results, expected)
 
 
+def test_agg_different_partials():
+    # for issue 28570
```
Comment (jbrockmendel) [inline pandas/tests/groupby/aggregate/test_aggregate.py:66]: nitpick: can you write this `GH#28570`?  makes it easier to grep for issue numbers

**Review by rhshadrach — CHANGES_REQUESTED** [review]
I don't think this closes the issue, see https://github.com/pandas-dev/pandas/issues/28570#issuecomment-1030697905 for details.

**Comment by mroeschke** [comment]
Thanks for the PR, but it appears to have gone stale. If you'd like to continue and investigate the bug fix, please ping and we can reopen.

**Pull request closed without merging**